### PR TITLE
Feature/bca/sso reauth

### PR DESCRIPTION
**Pull request by BillCarsonFr**
### Pull Request Checklist

This PR contains the following changes:
 - SSO UIA support
 - Design update for social logins button (provider brand + gitlab branding)
 - Align with web and iOS on e2e by default (relaxed some rules)
 - Ensure to setup cross signing if not yet done using server grace period if any (avoid non cross sign social logins)

## SDK changes for UIA

API calls (Tasks) that can trigger UIA flows now offers a new paramater 'UserInteractiveAuthInterceptor'
The task will call the interceptor to give him a chance to complete stages and provides an updated auth Dict.
The interceptor is given a 'promise' object, that the caller invokes when user inputs are collected.
Example:

````
awaitCallback<Unit> {
              session.cryptoService().deleteDevice(deviceId, object : UserInteractiveAuthInterceptor {
                  override fun performStage(flow: RegistrationFlowResponse, errorCode: String?, promise: Continuation<UIABaseAuth>) {
                      Timber.d("## UIA : deleteDevice UIA")
                      if (flow.nextUncompletedStage() == LoginFlowTypes.PASSWORD) {
                          [ ... PROMPT USER PASSWORD ] { pwd ->
                               // callback from dialog / fragment / activity
                              promise.resume(pwd)
                          }
                      } else {
                          // unsupported flow?
                          promise.resumeWith(Result.failure((IllegalArgumentException())))
                      }
                  }
              }, it)
          }

````

## App changes for UIA

New UIA flow (and support for UIA SSO)  implemented for:
- [x]  Upload cross signing keys
- [x]  Delete Device
- [x]  Add 3pid
- [x]  Deactivate

Introduction of a new ReAuthActivity that takes as param the UIA FlowResponse error and present needed UX to perform stage (for now supports password & SSO)

image

image

## Social Login buttons update

image

## e2e by default change summary

Previously, e2e was enabled by default if enabled by server admin, cross signing was enabled and trusted, no 3pid invites, and dest have some e2e capable devices

Now: Removed checks on cross signing and 3pid invites

### CODE REVIEW

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/failure/Extensions.kt [hunk 1]
```diff
@@ -53,6 +53,16 @@ fun Throwable.toRegistrationFlowResponse(): RegistrationFlowResponse? {
                     .adapter(RegistrationFlowResponse::class.java)
                     .fromJson(this.errorBody)
         }
+    } else if (this is Failure.ServerError && this.error.code == MatrixError.M_FORBIDDEN) {
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/failure/Extensions.kt:62]: You do not test the httpCode here? Maybe not mandatory and not specced though...

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/failure/MatrixError.kt [hunk 2]
```diff
@@ -42,7 +44,17 @@ data class MatrixError(
...
+
+        // For M_FORBIDDEN UIA
+        @Json(name = "session")
+        val session: String? = null,
+        @Json(name = "completed")
+        val completedStages: List<String>? = null,
+        @Json(name = "flows")
+        val flows: List<InteractiveAuthenticationFlow>? = null,
+        @Json(name = "params")
+        val params: JsonDict? = null
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/failure/MatrixError.kt:57]: That's really messy now :/

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/auth/data/LoginFlowResponse.kt [hunk 3]
```diff
@@ -43,5 +43,11 @@ internal data class LoginFlow(
...
         @Json(name = "org.matrix.msc2858.identity_providers")
-        val ssoIdentityProvider: List<SsoIdentityProvider>?
-)
+        val _devSsoIdentityProvider: List<SsoIdentityProvider>? = null,
+
+        @Json(name = "identity_providers")
+        val _ssoIdentityProvider: List<SsoIdentityProvider>? = null
+
+) {
+        val ssoIdentityProvider = _ssoIdentityProvider ?: _devSsoIdentityProvider
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/auth/data/LoginFlowResponse.kt:52]: You cannot do that if the MSC is not merged. Is it merged?

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/auth/registration/UIAExt.kt [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+import org.matrix.android.sdk.internal.crypto.model.rest.UIABaseAuth
+import timber.log.Timber
+import kotlin.coroutines.suspendCoroutine
+
+fun RegistrationFlowResponse.nextUncompletedStage(flowIndex: Int = 0): String? {
+    val completed = completedStages ?: emptyList()
+    return flows?.getOrNull(flowIndex)?.stages?.firstOrNull { completed.contains(it).not() }
+}
+
+suspend fun handleUIA(failure: Throwable, interceptor: UserInteractiveAuthInterceptor, retryBlock: suspend (UIABaseAuth) -> Unit): Boolean {
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/auth/registration/UIAExt.kt:31]: This should be internal

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/auth/registration/UIAExt.kt [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+package org.matrix.android.sdk.internal.auth.registration
+
+import org.matrix.android.sdk.api.auth.UserInteractiveAuthInterceptor
+import org.matrix.android.sdk.api.failure.Failure
+import org.matrix.android.sdk.api.failure.toRegistrationFlowResponse
+import org.matrix.android.sdk.internal.crypto.model.rest.UIABaseAuth
+import timber.log.Timber
+import kotlin.coroutines.suspendCoroutine
+
+fun RegistrationFlowResponse.nextUncompletedStage(flowIndex: Int = 0): String? {
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/auth/registration/UIAExt.kt:26]: This should be moved to the `api` package

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/model/rest/TokenBasedAuth.kt [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+import com.squareup.moshi.Json
+import com.squareup.moshi.JsonClass
+import org.matrix.android.sdk.api.auth.data.LoginFlowTypes
+
+/**
+ * This class provides the authentication data by using user and password
+ */
+@JsonClass(generateAdapter = true)
+data class TokenBasedAuth(
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/model/rest/TokenBasedAuth.kt:26]: I think there is a lot of missing `internal` keywords for this classes.

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/DefaultSession.kt [hunk 7]
```diff
@@ -273,6 +273,19 @@ internal class DefaultSession @Inject constructor(
...
     }
 
+    override fun getUIASsoFallbackUrl(authenticationSessionId: String): String {
+        val hsBas = sessionParams.homeServerConnectionConfig
+                .homeServerUri
+                .toString()
+                .trim { it == '/' }
+        return buildString {
+            append(hsBas)
+            append("/_matrix/client/r0/auth/m.login.sso/fallback/web")
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/DefaultSession.kt:283]: There is no tmp prefix here?

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/create/CreateRoomBodyBuilder.kt [hunk 8]
```diff
@@ -143,9 +143,11 @@ internal class CreateRoomBodyBuilder @Inject constructor(
...
 
     private suspend fun canEnableEncryption(params: CreateRoomParams): Boolean {
-        return (params.enableEncryptionIfInvitedUsersSupportIt
-                && crossSigningService.isCrossSigningVerified()
-                && params.invite3pids.isEmpty())
+        return params.enableEncryptionIfInvitedUsersSupportIt
+                // Parity with web, enable if users have encryption ready devices
+                // for now remove checks on cross signing and 3pid invites
+                // && crossSigningService.isCrossSigningVerified()
+                // && params.invite3pids.isEmpty())
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/create/CreateRoomBodyBuilder.kt:150]: What about the other checks below, for instance if a user has no device, encryption will not be enabled? Or if there are only 3pid invite?
This change should also probably be handled as a separate PR, but I now we are in a rush...

File: matrix-sdk-android/src/main/res/values/strings.xml [hunk 9]
```diff
@@ -305,4 +305,5 @@
 
     <string name="key_verification_request_fallback_message">%s is requesting to verify your key, but your client does not support in-chat key verification. You will need to use legacy key verification to verify keys.</string>
 
+    <string name="failed_to_initialize_cross_signing">Failed to set up Cross Signing</string>
```
Comment (bmarty) [inline matrix-sdk-android/src/main/res/values/strings.xml:308]: This is not the correct file to add this string

File: vector/src/main/java/im/vector/app/core/di/FragmentModule.kt [hunk 10]
```diff
@@ -522,8 +522,8 @@ interface FragmentModule {
 
     @Binds
     @IntoMap
-    @FragmentKey(BootstrapAccountPasswordFragment::class)
-    fun bindBootstrapAccountPasswordFragment(fragment: BootstrapAccountPasswordFragment): Fragment
+    @FragmentKey(BootstrapReAuthFragment::class)
+    fun bindBootstrapAccountPasswordFragment(fragment: BootstrapReAuthFragment): Fragment
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/core/di/FragmentModule.kt:526]: A detail, but the method should be renamed...

File: vector/src/main/java/im/vector/app/features/auth/ReAuthActivity.kt [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+        supportActionBar?.setTitle(title) ?: run { setTitle(title) }
+
+//        val authArgs = intent.getParcelableExtra<Args>(MvRx.KEY_ARG)
+
+        // For the sso flow we can for now only rely on the fallback flow, that handles all
+        // the UI, due to the sandbox nature of CCT (chrome custom tab) we cannot get much information
+        // on how the process did go :/
+        // so we assume that after the user close the tab we return success and let caller retry the UIA flow :/
+
+        addFragment(
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/auth/ReAuthActivity.kt:88]: Test if this is the first creation of the Activity? I think rotating the screen will add another Fragment

File: vector/src/main/java/im/vector/app/features/auth/ReAuthActivity.kt [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+                setResult(RESULT_OK, Intent().apply {
+                    putExtra(RESULT_FLOW_TYPE, LoginFlowTypes.SSO)
+                })
+                finish()
+            }
+        }
+    }
+
+    override fun onStart() = withState(sharedViewModel) { state ->
+        super.onStart()
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/auth/ReAuthActivity.kt:128]: weird to call `super` inside the `withState` block.

File: vector/src/main/java/im/vector/app/features/auth/ReAuthState.kt [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+) : MvRxState {
+    constructor(args: ReAuthActivity.Args) : this(
+            args.title,
+            args.session,
+            args.flowType,
+            lastErrorCode = args.lastErrorCode,
+            resultKeyStoreAlias = args.resultKeyStoreAlias
+    )
+
+    constructor() : this(null, null)
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/auth/ReAuthState.kt:38]: Is this constructor necessary?

File: vector/src/main/java/im/vector/app/features/crypto/recover/BootstrapBottomSheet.kt [hunk 14]
```diff
@@ -64,6 +68,25 @@ class BootstrapBottomSheet : VectorBaseBottomSheetDialogFragment<BottomSheetBoot
         return BottomSheetBootstrapBinding.inflate(inflater, container, false)
     }
 
+    private val reAuthActivityResultLauncher = registerStartForActivityResult { activityResult ->
+        if (activityResult.resultCode == Activity.RESULT_OK) {
+            when (activityResult.data?.extras?.getString(ReAuthActivity.RESULT_FLOW_TYPE)) {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/crypto/recover/BootstrapBottomSheet.kt:73]: Generally I prefer to keep the key for bundles private to the Activity and provide fun in the companion to get the result, which could be a sealed class for this case.

File: vector/src/main/java/im/vector/app/features/crypto/recover/BootstrapReAuthFragment.kt [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+    }
+
+    override fun invalidate() = withState(sharedViewModel) { state ->
+        if (state.step !is BootstrapStep.AccountReAuth) {
+            return@withState
+        }
+        val failure = state.step.failure
+        if (failure == null) {
+            views.reAuthFailureText.text = null
+            views.reAuthFailureText.isVisible = false
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/crypto/recover/BootstrapReAuthFragment.kt:74]: You can use `setTextOrHide()`

File: vector/src/main/java/im/vector/app/features/crypto/recover/BootstrapSharedViewModel.kt [hunk 16]
```diff
@@ -497,7 +510,7 @@ class BootstrapSharedViewModel @AssistedInject constructor(
                     }
                 }
             }
-            is BootstrapStep.SetupPassphrase                 -> {
+            is BootstrapStep.SetupPassphrase -> {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/crypto/recover/BootstrapSharedViewModel.kt:513]: Can you restore the formatting please?

File: vector/src/main/java/im/vector/app/features/settings/devices/DevicesViewModel.kt [hunk 17]
```diff
@@ -187,15 +202,45 @@ class DevicesViewModel @AssistedInject constructor(
...
-            is DevicesAction.Rename                 -> handleRename(action)
-            is DevicesAction.PromptRename           -> handlePromptRename(action)
-            is DevicesAction.VerifyMyDevice         -> handleInteractiveVerification(action)
-            is DevicesAction.CompleteSecurity       -> handleCompleteSecurity()
+            is DevicesAction.Refresh -> queryRefreshDevicesList()
+            is DevicesAction.Delete -> handleDelete(action)
+            is DevicesAction.Rename -> handleRename(action)
+            is DevicesAction.PromptRename -> handlePromptRename(action)
+            is DevicesAction.VerifyMyDevice -> handleInteractiveVerification(action)
+            is DevicesAction.CompleteSecurity -> handleCompleteSecurity()
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/settings/devices/DevicesViewModel.kt:210]: restore formatting?

File: vector/src/main/java/im/vector/app/features/settings/devices/VectorSettingsDevicesFragment.kt [hunk 18]
```diff
@@ -154,17 +150,34 @@ class VectorSettingsDevicesFragment @Inject constructor(
...
      */
-    private fun maybeShowDeleteDeviceWithPasswordDialog() {
-        if (mAccountPass.isNotEmpty()) {
-            viewModel.handle(DevicesAction.Password(mAccountPass))
-        } else {
-            PromptPasswordDialog().show(requireActivity()) { password ->
-                mAccountPass = password
-                viewModel.handle(DevicesAction.Password(mAccountPass))
-            }
+    private fun maybeShowDeleteDeviceWithPasswordDialog(reAuthReq: DevicesViewEvents.RequestReAuth) {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/settings/devices/VectorSettingsDevicesFragment.kt:175]: This method could be renamed, also to doc above is not correct anymore

File: vector/src/main/res/layout/fragment_bootstrap_reauth.xml [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+    <TextView
+        android:id="@+id/bootstrapDescriptionText"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:layout_marginTop="8dp"
+        android:textColor="?riotx_text_primary"
+        android:textSize="14sp"
+        app:layout_constraintBottom_toTopOf="@id/waitingProgress"
+        app:layout_constraintTop_toTopOf="parent"
+        tools:text="@string/re_authentication_activity_title" />
```
Comment (bmarty) [inline vector/src/main/res/layout/fragment_bootstrap_reauth.xml:19]: Would be better to use Epoxy here, but OK for a quicker implementation

File: vector/src/main/res/layout/fragment_reauth_confirm.xml [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+            app:layout_constraintTop_toTopOf="parent"
+            app:layout_constraintVertical_bias="0"
+            app:layout_constraintVertical_chainStyle="packed" />
+
+        <FrameLayout
+            android:id="@+id/passwordContainer"
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:layout_margin="16dp"
+            app:layout_constraintBottom_toTopOf="@id/loginPasswordNotice"
```
Comment (bmarty) [inline vector/src/main/res/layout/fragment_reauth_confirm.xml:32]: This constraint is in error, can be removed.

**Review by bmarty — CHANGES_REQUESTED** [review]
First static review.

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/auth/UserInteractiveAuthInterceptor.kt [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+ */
+interface UserInteractiveAuthInterceptor {
+
+    /**
+     * When the API needs additional auth, this will be called.
+     * Implementation should check the flows from flow response and act accordingly.
+     * Updated auth should be provided using promise.resume, this allow implementation to perform
+     * an async operation (prompt for user password, open sso fallback) and then resume initial API call when done.
+     */
+    fun performStage(flowResponse: RegistrationFlowResponse, errCode: String?, promise: Continuation<UIABaseAuth>)
```
Comment (ganfra) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/auth/UserInteractiveAuthInterceptor.kt:47]: This API is really weird to use :D

Comment (ganfra) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/auth/UserInteractiveAuthInterceptor.kt:47]: Other than that, got warnings because implementations haven't the same names for params (flow, errorCode)

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/Session.kt [hunk 22]
```diff
@@ -245,6 +245,8 @@ interface Session :
 
     val sharedSecretStorageService: SharedSecretStorageService
 
+    fun getUIASsoFallbackUrl(authenticationSessionId: String): String
```
Comment (ganfra) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/Session.kt:248]: Why is it directly available from the session, shouldn't be in some auth service ?

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/model/rest/TokenBasedAuth.kt (hunk 6 above)
Comment (ganfra) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/model/rest/TokenBasedAuth.kt:26]: Or it's public and on the wrong package

File: vector/src/main/java/im/vector/app/features/auth/ReAuthActivity.kt (hunk 12 above)
Comment (ganfra) [reply vector/src/main/java/im/vector/app/features/auth/ReAuthActivity.kt:128]: +1

File: vector/src/main/java/im/vector/app/features/auth/ReAuthActivity.kt (hunk 11 above)
Comment (ganfra) [reply vector/src/main/java/im/vector/app/features/auth/ReAuthActivity.kt:88]: +1

**Review by ganfra — CHANGES_REQUESTED** [review]
Some remarks
Also, tests didn't compile

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/failure/Extensions.kt (hunk 1 above)
Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/failure/Extensions.kt:62]: Yes it should be 401, i can add the check
> If the homeserver decides that an attempt on a stage was unsuccessful, but the client may make a second attempt, it returns the same HTTP status 401 response as above, with the addition of the standard errcode and error fields describing the error.

**Comment by bmarty** [comment]
Problem during testing:
- [ ] If I create an account and then deactivate it, I'm not asked to provide credential again. There should be no grace period to deactivate the account.
- [ ] If I try to do SSO with a deactivate account, I'm stuck to the "Your account has been deactivated" webpage forever (i.e. even if I kill the app)

File: vector/src/main/java/im/vector/app/core/platform/VectorBaseFragment.kt [hunk 23]
```diff
@@ -200,6 +200,7 @@ abstract class VectorBaseFragment<VB: ViewBinding> : BaseMvRxFragment(), HasScre
     }
 
     protected fun showLoadingDialog(message: CharSequence? = null, cancelable: Boolean = false) {
+        progress?.dismiss()
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/core/platform/VectorBaseFragment.kt:203]: We should probably do nothing if `progress` is not null here to avoid blink effect.

File: vector/src/main/java/im/vector/app/features/crypto/recover/BootstrapReAuthFragment.kt [hunk 24]
```diff
@@ -70,14 +71,12 @@ class BootstrapReAuthFragment @Inject constructor(
...
-            views.reAuthFailureText.text = null
-            views.reAuthFailureText.isVisible = false
+            views.reAuthFailureText.setTextOrHide(null)
             views.waitingProgress.isVisible = true
             views.bootstrapCancelButton.isVisible = false
             views.bootstrapRetryButton.isVisible = false
         } else {
-            views.reAuthFailureText.text = failure
-            views.reAuthFailureText.isVisible = true
+            views.reAuthFailureText.setTextOrHide(failure)
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/crypto/recover/BootstrapReAuthFragment.kt:79]: This could be done above the if block then :p

Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/crypto/recover/BootstrapReAuthFragment.kt:79]: This could be done above the if block then :p

**Review by bmarty — COMMENTED** [review]
Thanks for the update. 2 minor remarks.
Also there is lots of duplicated code now, we should probably try to improve that in the future.

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/DefaultSession.kt (hunk 7 above)
Comment (bmarty) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/DefaultSession.kt:283]: it's an old thing

File: vector/src/main/java/im/vector/app/features/crypto/recover/BootstrapBottomSheet.kt (hunk 14 above)
Comment (bmarty) [reply vector/src/main/java/im/vector/app/features/crypto/recover/BootstrapBottomSheet.kt:73]: To be reworked later

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/Session.kt (hunk 22 above)
Comment (bmarty) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/Session.kt:248]: It's OK for now

File: vector/src/main/java/im/vector/app/features/crypto/recover/BootstrapReAuthFragment.kt (hunk 24 above)
Comment (bmarty) [reply vector/src/main/java/im/vector/app/features/crypto/recover/BootstrapReAuthFragment.kt:79]: I do it

Comment (bmarty) [reply vector/src/main/java/im/vector/app/features/crypto/recover/BootstrapReAuthFragment.kt:79]: done

File: vector/src/main/java/im/vector/app/core/platform/VectorBaseFragment.kt (hunk 23 above)
Comment (bmarty) [reply vector/src/main/java/im/vector/app/core/platform/VectorBaseFragment.kt:203]: Ok, the text can change... Ok as a quick fix

**Review by bmarty — APPROVED** [review]
LGTM
